Sub: Enable Visual Odometry interface

Williangalvani opened the pull request:
This will enable SUB to use DVL (Doppler Velocity Logger) sensors for position control.
We have a blog post explaining it [here](https://bluerobotics.com/ardusub-research-development-dvl/).

The DVL outputs are very precise (usually < 1% error) and the output is quite compatible with the [VISION_POSITION_DELTA](https://mavlink.io/en/messages/ardupilotmega.html#VISION_POSITION_DELTA) message. They usually provides Z position, XY velocity (which we integrate) and most have a built-in IMU for attitude.

rmackay9 commented on ArduSub/Parameters.cpp:
```diff
@@ -536,6 +536,10 @@ const AP_Param::Info Sub::var_info[] = {
     GOBJECT(rally,      "RALLY_",   AP_Rally),
 #endif
 
+    // @Group: VISO
```
This probably needs to be surrounded by checks of VISUAL_ODOMETRY_ENABLED

rmackay9 commented on ArduSub/GCS_Sub.cpp:
```diff
@@ -66,7 +72,9 @@ void GCS_Sub::update_vehicle_sensor_status_flags()
     }
 #endif
-
+    if (visual_odom && visual_odom->enabled() && visual_odom->healthy()) {
```
I wonder if this also needs to be surrounded by checks of VISUAL_ODOMETRY_ENABLED

rmackay9 commented on ArduSub/control_poshold.cpp:
```diff
@@ -10,7 +10,7 @@
 {
     // fail to initialise PosHold mode if no GPS lock
-    if (!position_ok()) {
+    if (!position_ok() && (!visual_odom.enabled() || !visual_odom.healthy())) {
```
I'm surprised that this direct check of visual odometry is required.  If the EKF is happy and knows it's position then I think the existing position_ok() check should be sufficient.

Williangalvani replied:
I'm trying without it and I just can't get it into position hold. Did I miss something?
position_ok() for [sub](https://github.com/ArduPilot/ardupilot/blob/master/ArduSub/system.cpp#L224) seems to be the same as for [copter](https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/system.cpp#L296)

rmackay9 replied:
Maybe the EKF origin hasn't been set?  If using MP then try right-mouse-button-clicking on the Data screen's map and Set Home >> Set EKF Origin and see if it shows up.  After that I suspect it will work.

peterbarker commented on ArduSub/GCS_Sub.cpp:
```diff
@@ -36,7 +36,13 @@ void GCS_Sub::update_vehicle_sensor_status_flags()
     }
 #endif
-
+#if VISUAL_ODOMETRY_ENABLED == ENABLED
```
This should be moved up to the parent class and removed from Copter and Rover.

rmackay9 commented:
@Williangalvani,

FYI, I'm also working on the visual odometry library with this PR:https://github.com/ArduPilot/ardupilot/pull/13982.  My PR includes a small timing improvement to the handling of the position-delta-estimate message so if possible I'd like to get your help testing that my PR doesn't break that.

rmackay9 commented:
@Williangalvani,

In my PR: https://github.com/ArduPilot/ardupilot/pull/13982 I've moved AP_VisualOdom to the AP_Vehicle class so that all vehicles should be able to use it.  Also I think I've discovered why you needed to modify the position_ok() function and fixed it in an alternative way.

Do you think you could try testing my PR's branch to see if it works for you?  If not I'm happy to resolve any issues you face.

Williangalvani commented:
> @Williangalvani,
> 
> In my PR: #13982 I've moved AP_VisualOdom to the AP_Vehicle class so that all vehicles should be able to use it. Also I think I've discovered why you needed to modify the position_ok() function and fixed it in an alternative way.
> 
> Do you think you could try testing my PR's branch to see if it works for you? If not I'm happy to resolve any issues you face.

Thank you @rmackay9 !
I will take a look and test it today.

Williangalvani commented:
closing in favor of #13982
